// @guildedjs/common/dist/util.js
"use strict";
/* istanbul ignore file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ROUTES = exports.IMG_EXTENSION = void 0;
var IMG_EXTENSION;
(function (IMG_EXTENSION) {
    IMG_EXTENSION["PNG"] = "png";
    IMG_EXTENSION["GIF"] = "gif";
    IMG_EXTENSION["WEBP"] = "webp";
})(IMG_EXTENSION = exports.IMG_EXTENSION || (exports.IMG_EXTENSION = {}));
const formAssetURL = (route, hash, extension, width, height, size) => {
    const url = new URL(`https://${exports.ROUTES.IMAGE_CDN_DOMAIN}/${route}/${hash}-${size}.${extension.toLowerCase()}`);
    if (width)
        url.searchParams.append('w', width.toString());
    if (height)
        url.searchParams.append('h', height.toString());
    return url.toString();
};
exports.ROUTES = {
    ASSETS: {
        AVATAR_URL: (hash, size = 'Medium') => formAssetURL('UserAvatar', hash, IMG_EXTENSION.PNG, undefined, undefined, size),
        IMAGE_IN_CHAT: (hash, size = 'Full', width, height) => formAssetURL('ContentMedia', hash, IMG_EXTENSION.WEBP, width, height, size),
        PROFILE_BANNER: (hash, size = 'Hero', width, height) => formAssetURL('UserBanner', hash, IMG_EXTENSION.PNG, width, height, size),
        TEAM_BANNER: (hash, size = 'Hero', width, height) => formAssetURL('TeamBanner', hash, IMG_EXTENSION.PNG, width, height, size),
        TEAM_EMOJI: (hash, size = 'Full', extension = 'WEBP', width, height) => formAssetURL('CustomReaction', hash, extension.toLowerCase(), width, height, size),
        TEAM_ICON: (hash, size = 'Medium') => formAssetURL('TeamAvatar', hash, IMG_EXTENSION.PNG, undefined, undefined, size),
    },
    AWS_CDN: 'https://s3-us-west-2.amazonaws.com/www.guilded.gg/',
    BASE_DOMAIN: 'api.guilded.gg',
    CDN: 'img.guildedcdn.com',
    IMAGE_CDN_DOMAIN: 'img.guildedcdn.com',
    MEDIA_DOMAIN: 'media.guilded.gg',
};
