// @guildedjs/guilded.js/dist/structures/managers/TeamGroupChannelManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamGroupChannelManager = void 0;
const Channel_1 = require("../Channel");
const BaseManager_1 = require("./BaseManager");
class TeamGroupChannelManager extends BaseManager_1.BaseManager {
    constructor(client, group) {
        var _a, _b, _c;
        super(client, Channel_1.TeamChannel, { maxSize: (_c = (_b = (_a = client.options) === null || _a === void 0 ? void 0 : _a.cache) === null || _b === void 0 ? void 0 : _b.cacheMaxSize) === null || _c === void 0 ? void 0 : _c.channelsCache });
        this.group = group;
    }
    /**
     * Create a channel in the group this manager belongs to.
     */
    create({ name, parent, type, isPublic = true, }) {
        var _a;
        return this.client.rest.post(`/teams/${(_a = this.group.team) === null || _a === void 0 ? void 0 : _a.id}/groups/${this.group.id}/channels`, {
            channelCategoryId: parent,
            contentType: type,
            isPublic: isPublic,
            name,
        });
    }
}
exports.TeamGroupChannelManager = TeamGroupChannelManager;
