// @guildedjs/guilded.js/dist/ws/Heartbeater.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Heartbeater {
    constructor(gateway) {
        this.gateway = gateway;
        this.lastPingSentAt = 0;
        this.active = false;
    }
    start(interval) {
        var _a, _b, _c;
        this.gateway.client.debug('Heartbeating...');
        this.active = true;
        this.interval = setInterval(this.sendHB.bind(this), (_c = interval !== null && interval !== void 0 ? interval : (_b = (_a = this.gateway.client.options) === null || _a === void 0 ? void 0 : _a.ws) === null || _b === void 0 ? void 0 : _b.heartbeatInterval) !== null && _c !== void 0 ? _c : 15000);
    }
    destroy() {
        clearInterval(this.interval);
        this.active = false;
    }
    sendHB() {
        if (!this.gateway.ws)
            return;
        if (this.gateway.ws.readyState !== 1)
            return;
        this.lastPingSentAt = Date.now();
        this.gateway.ws.send('2');
        this.gateway.client.rest.put('/users/me/ping');
        this.gateway.client.debug('heartbeat sent');
    }
}
exports.default = Heartbeater;
