// @guildedjs/guilded.js/dist/ws/events/ChatMessageCreated.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const structures_1 = require("../../structures");
const Message_1 = require("../../structures/Message");
const typings_1 = require("../../typings");
const Event_1 = __importDefault(require("./Event"));
class ChatMessageCreatedEvent extends Event_1.default {
    constructor(client) {
        super(client);
    }
    ingest(data) {
        var _a;
        if (data) {
            let channel = this.client.channels.cache.get(data.channelId);
            const team = (_a = (data.teamId ? this.client.teams.cache.get(data.teamId) : null)) !== null && _a !== void 0 ? _a : null;
            if (!channel) {
                channel = new structures_1.PartialChannel(this.client, {
                    contentType: data.contentType,
                    createdAt: data.createdAt,
                    createdBy: data.createdBy,
                    id: data.channelId,
                    type: data.channelType,
                }, team);
                this.client.channels.cache.set(channel.id, channel);
            }
            const newMessage = new Message_1.Message(this.client, Object.assign({ channelId: data.channelId, teamId: data.teamId }, data.message), channel);
            channel.messages.cache.set(newMessage.id, newMessage);
            this.client.emit(typings_1.events.MESSAGE_CREATE, newMessage);
            return [true];
        }
        return [false, 'passthrough'];
    }
}
exports.default = ChatMessageCreatedEvent;
