// @guildedjs/guilded.js/dist/ws/events/ChatMessageReactionAdded.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const structures_1 = require("../../structures");
const typings_1 = require("../../typings");
const Event_1 = __importDefault(require("./Event"));
class ChatMessageReactionAddedEvent extends Event_1.default {
    constructor(client) {
        super(client);
    }
    ingest(data) {
        var _a, _b, _c, _d, _e, _f, _g;
        if (data) {
            const reacter = this.client.users.cache.get(data.reaction.createdBy);
            let channel = this.client.channels.cache.get(data.channelId);
            if (!channel) {
                if (!((_b = (_a = this.client.options) === null || _a === void 0 ? void 0 : _a.partials) === null || _b === void 0 ? void 0 : _b.includes('CHANNEL')))
                    return [false, 'Uncached channel'];
                channel = this.client.channels.add({
                    contentType: data.contentType,
                    id: data.channelId,
                    teamId: (_c = data.teamId) !== null && _c !== void 0 ? _c : undefined,
                    type: data.channelType,
                });
            }
            const message = (_d = channel === null || channel === void 0 ? void 0 : channel.messages) === null || _d === void 0 ? void 0 : _d.cache.get(data.message.id);
            if (!message && !((_f = (_e = this.client.options) === null || _e === void 0 ? void 0 : _e.partials) === null || _f === void 0 ? void 0 : _f.includes('MESSAGE')))
                return [false, 'Uncached message'];
            let messageReaction = message === null || message === void 0 ? void 0 : message.reactions.get(data.reaction.customReactionId.toString());
            if (!messageReaction) {
                const newMessageReaction = new structures_1.MessageReaction(this.client, Object.assign(Object.assign({}, data.reaction), { createdAt: new Date().toISOString(), users: [{ botId: null, id: data.reaction.createdBy, webhookId: null }] }));
                message === null || message === void 0 ? void 0 : message.reactions.set(newMessageReaction.id, newMessageReaction);
                messageReaction = newMessageReaction;
            }
            messageReaction.users.set(data.reaction.createdBy, (_g = this.client.users.cache.get(data.reaction.createdBy)) !== null && _g !== void 0 ? _g : { id: data.reaction.createdBy });
            this.client.emit(typings_1.events.MESSAGE_REACTION_ADD, messageReaction, reacter !== null && reacter !== void 0 ? reacter : data.reaction.createdBy);
            return [true];
        }
        return [false, 'passthrough'];
    }
}
exports.default = ChatMessageReactionAddedEvent;
